A3-channel-tracing.md

```diff
@@ -0,0 +1,204 @@
+   // tracks total trace nodes seen, since many may be overwritten
+  int64 num_nodes_logged = 2;
+  // time for creation for the channel or subchannel tracer
+  google.protobuf.Timestamp start_time = 3;
+  // all of the data for this tracer
+  repeated TraceNode nodes = 4;
+}
+
+// trace for one channel
+message ChannelTrace {
+  // parent channel data
+  TraceData channel_data = 1;
+  // data for the subchannels of this channel
+  repeated TraceData subchannel_data = 3;
+}
+```
+
+## Rationale
+
+This approach makes it easy to add trace, since each tracer is attached to its respective channel or subchannel. The tracer will also be passed down into the resolver and load balancer to record trace for interesting events that happen there. 
+
+One downside to this approach is the potential for the tracer to bloat memory. For this reason, there will be a new channel argument that sets a limit for how many trace nodes each channel or subchannel will hold. This is described in more detail in the [Garbage Collection](#garbage-collection) section.
+
+## Implementation
+
+### Basic Design
+
+The implementation will be done in the C-core first, with Java and Go to follow. The tracer object will hold a linked list of trace nodes. If a trace node pertains to a subchannel, then it will point to a tracer for that subchannel. tracers are refcounted objects.
+
+```C
+// One node of tracing data
+struct grpc_trace_node {
+  const char* data;
+  grpc_error* error;
+  gpr_timespec time_created;
+  grpc_connectivity_state connectivity_state;
+  grpc_trace_node* next;
+
+  // The tracer object that owns this trace node. This is used to ref and
+  // unref the tracing object as nodes are added or overwritten
+  grpc_channel_tracer* subchannel;
+};
+
+struct grpc_trace_node_list {
+  size_t size;
+  size_t max_size;
+  grpc_trace_node* head_trace;
+  grpc_trace_node* tail_trace;
+};
+
+/* the channel tracing object */
+struct grpc_channel_tracer {
+  gpr_refcount refs;
+  gpr_mu tracer_mu; 
+  int64_t num_nodes_logged;
+  grpc_trace_node_list node_list;
+  gpr_timespec time_created;
+};
+```
+
+### Tracer Functions
+
+The tracer files with expose several calls which with the C-core code can interact with tracers:
+
+```C
+/* Initializes the tracing object with gpr_malloc. The caller has
+   ownership over the returned tracing object */
+grpc_channel_tracer* grpc_channel_tracer_create(size_t max_nodes);
+
+/* Adds a new trace node to the tracing object */
+void grpc_channel_tracer_add_trace(grpc_channel_tracer* tracer, 
+    char* trace, struct grpc_error* error,
+    grpc_connectivity_state connectivity_state);
+
+/* Frees all of the resources held by the tracer object */
+void grpc_channel_tracer_unref(grpc_channel_tracer* tracer);
+```
+### API Changes
+
+And finally, a new method will be added to the channel surface:
+
+```C
+/* returns the tracing data in the form of a grpc json object */
+grpc_json* grpc_channel_get_trace(grpc_channel* channel);
```

Comment:
I'm not sure that we want to expose the `grpc_json` type into the C-core public API.  It might be better to return a JSON string instead of the parsed JSON tree.